### PR TITLE
Tls compressed points

### PR DESCRIPTION
**Pull request by cordney**
Adds support for the TLS Supported Point Format Extension per RFC 4492. Users can control compressed/uncompressed via a new `TLS_Policy::use_ecc_point_compression()` toggle. If compression is preferred by either party, client kex and server kex send the ECDH public point as compressed, respectively. Note that per RFC 4492 also client and server certificate messages have to respect the preference of the other party, but we leave this to the user and currently don't check the certificate.

First part to implement #636.

### CODE REVIEW

File: src/lib/tls/msg_server_kex.cpp [hunk 1]
```diff
@@ -85,7 +85,15 @@ Server_Key_Exchange::Server_Key_Exchange(Handshake_IO& io,
...
 
-      append_tls_length_value(m_params, ecdh->public_value(), 1);
+      // follow client's preference for point compression
+      if(state.client_hello()->prefers_compressed_ec_points())
+         {
+         append_tls_length_value(m_params, ecdh->public_value(PointGFp::COMPRESSED), 1);
+         }
+      else
+         {
+         append_tls_length_value(m_params, ecdh->public_value(PointGFp::UNCOMPRESSED), 1);
```
Comment (randombit) [inline src/lib/tls/msg_server_kex.cpp]: Same ?:

File: src/lib/tls/tls_extensions.cpp [hunk 2]
```diff
@@ -353,6 +356,51 @@ Supported_Elliptic_Curves::Supported_Elliptic_Curves(TLS_Data_Reader& reader,
...
+
+   // if we send this extension, we prefer compressed points,
+   // otherwise we don't send it (which is equal to supporting only uncompressed)
+   buf.push_back(ANSIX962_COMPRESSED_PRIME);
+
+   // if this extension is sent, it MUST include uncompressed (RFC 4492, section 5.1)
+   buf.push_back(UNCOMPRESSED);
+
+   buf[0] = static_cast<byte>(buf.size()-1);
+
```
Comment (randombit) [inline src/lib/tls/tls_extensions.cpp]: Why not `return std::vector<byte>{2, ANSI_X962_COMPRESSED_PRIME, UNCOMPRESSED}` ?

Comment (cordney) [inline src/lib/tls/tls_extensions.cpp]: Nice one, thx.

File: src/lib/tls/msg_server_kex.cpp (hunk 1 above)
Comment (cordney) [inline src/lib/tls/msg_server_kex.cpp]: Fixed

File: src/lib/tls/msg_client_kex.cpp [hunk 3]
```diff
@@ -172,7 +172,15 @@ Client_Key_Exchange::Client_Key_Exchange(Handshake_IO& io,
...
 
-         append_tls_length_value(m_key_material, priv_key.public_value(), 1);
+         // follow server's preference for point compression
+         if(state.server_hello()->prefers_compressed_ec_points())
+            {
+            append_tls_length_value(m_key_material, priv_key.public_value(PointGFp::COMPRESSED), 1);
+            }
+         else
+            {
+            append_tls_length_value(m_key_material, priv_key.public_value(PointGFp::UNCOMPRESSED), 1);
```
Comment (cordney) [inline src/lib/tls/msg_client_kex.cpp]: Fixed

**Comment by cordney** [comment]
Also fixed the tls_messages tests. Somehow this passed unnoticed locally.

**Comment by cordney** [comment]
Strange test failure on cross-arm32:

```
FFI ran 226 tests 1 FAILED
Failure 1: FFI botan_privkey_export_encrypted unexpectedly failed with error code -10
```

**Comment by randombit** [comment]
I've been chasing that FFI test error for a while and thought I had fixed it in 4e42b0186b but it still recurs. It is not ARM specific but does seem to be related to how slow the machine is during the test run - it is much more commonly seen in QEMU or AppVeyor builds. I have never been able to repro it locally.

**Pull request merged**